reactor-netty-http/src/test/java/reactor/netty/http/HttpMetricsHandlerTests.java

```diff
@@ -1174,4 +1211,128 @@ public void recordConnectTime(SocketAddress socketAddress, Duration duration, St
+		final static ResponseSentHandler INSTANCE = new ResponseSentHandler();
+		AtomicReference<CountDownLatch> latchRef;
+
+		void register(AtomicReference<CountDownLatch> latchRef, ChannelPipeline pipeline) {
+			this.latchRef = latchRef;
+			pipeline.addBefore(NettyPipeline.HttpMetricsHandler, HANDLER_NAME, this);
+		}
+
+		void register(CountDownLatch latch, ChannelPipeline pipeline) {
+			register(new AtomicReference<>(latch), pipeline);
+		}
+
+		@Override
+		@SuppressWarnings("FutureReturnValueIgnored")
+		public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) {
+			if (msg instanceof LastHttpContent) {
+				promise.addListener(future -> {
+					latchRef.get().countDown();
+				});
+			}
+
+			ctx.write(msg, promise);
+		}
+
+		@Override
+		public boolean isSharable() {
+			return true; // A server may accept multiple connections, hence this handler must be sharable
+		}
+	}
+
+	/**
+	 * Server Handler used to detect when the last http client request content has been received by the server.
+	 * Handler placed after the HttpMetricsHandler on the Server pipeline.
+	 * Metrics are up-to-date when the latch is counted down.
+	 */
+	static final class RequestReceivedHandler extends ChannelInboundHandlerAdapter {
+		final static RequestReceivedHandler INSTANCE = new RequestReceivedHandler();
+		final static String HANDLER_NAME = "ServerReceivedHandler.handler";
+		AtomicReference<CountDownLatch> latchRef;
+
+		void register(AtomicReference<CountDownLatch> latchRef, ChannelPipeline pipeline) {
+			this.latchRef = latchRef;
+			pipeline.addAfter(NettyPipeline.HttpMetricsHandler, HANDLER_NAME, this);
+		}
+
+		void register(CountDownLatch latch, ChannelPipeline pipeline) {
+			register(new AtomicReference<>(latch), pipeline);
+		}
+
+		@Override
+		public void channelRead(ChannelHandlerContext ctx, Object msg) {
+			if (msg instanceof LastHttpContent) {
+				latchRef.get().countDown();
+			}
+			ctx.fireChannelRead(msg);
+		}
+
+		@Override
+		public boolean isSharable() {
+			return true;
+		}
+	}
+
+	/**
+	 * Server handler used to wait until the client socket is closed on the server side.
+	 * For HTTP1.1, the handler is placed before the ReactorBridge, so all previous handlers will see
+	 * the close before this handler. For HTTP2, the handler is placed lastly on the pipeline.
+	 */
+	static final class ServerCloseHandler extends ChannelInboundHandlerAdapter {
+		static final ServerCloseHandler INSTANCE = new ServerCloseHandler();
+		static final String HANDLER_NAME = "ServerCloseHandler.handler";
+		private CountDownLatch latch;
+
+		void register(Channel channel, CountDownLatch latch, boolean http11) {
+			this.latch = latch;
+
+			if (http11) {
+				channel.pipeline().addBefore(NettyPipeline.ReactiveBridge, HANDLER_NAME, this);
+			}
+			else {
+				channel.parent().pipeline().addLast(HANDLER_NAME, this);
+			}
+		}
+
+		@Override
+		public void channelInactive(ChannelHandlerContext ctx) throws Exception {
```

Comment:
`Exception` is never thrown
```suggestion
		public void channelInactive(ChannelHandlerContext ctx) {
```